### PR TITLE
doc: add style guide

### PR DESCRIPTION
**Pull request by black-sliver**

### CODE REVIEW

**Review by Berserker66 — APPROVED** [review]

**Comment by Berserker66** [comment]
This might be one where we should wait for a few more eyes and approvals.

**Review by ThePhar — APPROVED** [review]
I'm on board. Do have one question though.

Is this style guide also supposed to be enforced for world definitions? That's the only place I can see some friction come in.

File: docs/style.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# Style Guide
+
+## Generic
+
+* This guide can be ignored for data files that are not to be viewed in an editor.
+* 120 character per line for all source files
+* Avoid white space errors like trailing spaces.
```
Comment (Ijwu) [inline docs/style.md:7]: Define all whitespace errors to avoid or point me to the section of PEP8 where this is relevant.

(If we're giving people guidelines, I think they should be specific and actionable.)

File: docs/style.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
+# Style Guide
+
+## Generic
+
+* This guide can be ignored for data files that are not to be viewed in an editor.
+* 120 character per line for all source files
```
Comment (Ijwu) [inline docs/style.md:6]:
```suggestion
* 120 character per line for all source files.
```

File: docs/style.md [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+* Avoid white space errors like trailing spaces.
+
+
+## Python Code
+
+* We mostly follow [PEP8](https://peps.python.org/pep-0008/). Read below to see the differences.
+* 120 characters per line. PyCharm does this automatically, other editors can be configured for it.
+* Strings in core code will be `"strings"`
```
Comment (Ijwu) [inline docs/style.md:14]:
```suggestion
* Strings in core code will be `"strings"`. In other words: double quote your strings.
```

File: docs/style.md [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+* Use type annotation where possible
```
Comment (Ijwu) [inline docs/style.md:15]:
```suggestion
* Use type annotation where possible.
```

File: docs/style.md [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+
+## Markdown
+
+* We almost follow [Google's styleguide](https://google.github.io/styleguide/docguide/style.html).
+  Read below for differences.
+* For existing documents, try to follow its style or ask to completely reformat it.
+* 120 characters per line
```
Comment (Ijwu) [inline docs/style.md:23]:
```suggestion
* 120 characters per line.
```

File: docs/style.md [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+* One space between bullet/number and text
```
Comment (Ijwu) [inline docs/style.md:24]:
```suggestion
* One space between bullet/number and text.
```

File: docs/style.md [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+* No lazy numbering
```
Comment (Ijwu) [inline docs/style.md:25]: I assume this means don't do:
```md
1. item 1
1. item 2
1. item 3
```
Is that a correct assumption?

```suggestion
* No lazy numbering.
```

**Review by Ijwu — CHANGES_REQUESTED** [review]
@LegendaryLinux would you please provide a JS style guide? If you feel it's applicable, HTML/CSS as well.

File: docs/style.md (hunk 7 above)
Comment (black-sliver) [reply docs/style.md:25]: Yes. Files we currently have do not number lazily and, unless you are very cautious with the other formatting, lazy numbering may be very hard to follow in code or when comparing render to code.

File: docs/style.md (hunk 1 above)
Comment (black-sliver) [reply docs/style.md:7]: My idea was to put it in the generic section and let PEP8 (linked below) and the Markdown guide (also linked below) handle the rest; providing a full list here may sound like it's overruling what's defined in the links below, but trailing whitespace applies to basically all languages, so i thought it might be a good example.

**Comment by black-sliver** [comment]
> Is this style guide also supposed to be enforced for world definitions? That's the only place I can see some friction come in.

I think what is being "enforced" already as soon as ~~berserker~~ PyCharm touches the code is PEP8 with 120 chars, minus the string recommendation. Regarding strings your world may be odd because it copies code with single quotes from another project, but i think double quotes would be preferred in other cases- we could define that somehow in the guide, maybe.

**Comment by Ijwu** [comment]
> > Is this style guide also supposed to be enforced for world definitions? That's the only place I can see some friction come in.
> 
> I think what is being "enforced" already as soon as ~berserker~ PyCharm touches the code is PEP8 with 120 chars, minus the string recommendation. Regarding strings: your world may be odd because it copies code with single quotes from another project, but i think double quotes would be preferred in other cases- we could define that somehow in the guide, maybe.

I think we just need to make the decision whether or not worlds have to follow the style guide. I would vote yes, as all code should generally do so within AP. We can't have "pick-and-choose" standards, in my opinion.

File: docs/style.md (hunk 1 above)
Comment (black-sliver) [reply docs/style.md:7]: This is a bit esoteric, but
* we don't want white space errors in any source file
* PEP8 only talks about python
* Markdown styleguide only talks about markdown
* Trailing whitespace is a common mistake that gets highlighted by git and github diffs

The only other common thing that comes to mind is the number of newlines at end of file.

**Comment by black-sliver** [comment]
docs/style.md should apply to everything that is in this repo. If there are exceptions, they should be noted here. To me this feels implied if we talk about core strings and world strings in the text already.

File: docs/style.md [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+* This guide can be ignored for data files that are not to be viewed in an editor.
+* 120 character per line for all source files.
+* Avoid white space errors like trailing spaces.
+
+
+## Python Code
+
+* We mostly follow [PEP8](https://peps.python.org/pep-0008/). Read below to see the differences.
+* 120 characters per line. PyCharm does this automatically, other editors can be configured for it.
+* Strings in core code will be `"strings"`. In other words: double quote your strings.
```
Comment (Jarno458) [inline docs/style.md:14]: Why enforce double quotes? if a string uses double or single quote, your editor will color it as string, doesnt matter which veriant you use

File: docs/style.md [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+
+* This guide can be ignored for data files that are not to be viewed in an editor.
+* 120 character per line for all source files.
+* Avoid white space errors like trailing spaces.
+
+
+## Python Code
+
+* We mostly follow [PEP8](https://peps.python.org/pep-0008/). Read below to see the differences.
+* 120 characters per line. PyCharm does this automatically, other editors can be configured for it.
```
Comment (Jarno458) [inline docs/style.md:13]: i hate limiting line lengths, its more importent to create good readable code. even if some parameters are not directly visable on screen. it can still be better readable then splitng a function call over multiple lines

Also 120 is a really old value people used when they resolutuions of max 1024px

File: docs/style.md [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+
+
+## Markdown
+
+* We almost follow [Google's styleguide](https://google.github.io/styleguide/docguide/style.html).
+  Read below for differences.
+* For existing documents, try to follow its style or ask to completely reformat it.
+* 120 characters per line.
+* One space between bullet/number and text.
+* No lazy numbering.
```
Comment (Jarno458) [inline docs/style.md:28]: Why not? this is a feature that makes editing easyer.

**Review by Jarno458 — CHANGES_REQUESTED** [review]

File: docs/style.md (hunk 9 above)
Comment (black-sliver) [reply docs/style.md:13]: 120 will make two code windows fit side-by-side on modern screens. 80 was the old one.

File: docs/style.md (hunk 8 above)
Comment (black-sliver) [reply docs/style.md:14]: To make the **core** code consistent. This was discussed in discord and the majority was for using `"`

File: docs/style.md (hunk 10 above)
Comment (black-sliver) [reply docs/style.md:28]: Couple of reasons:
1. Lazy numbering is harder to read in source form unless you use more space(s) than we do
1. which means harder to modify/correlate to the render
1. Lazy numbering shines where a list gets longer than what is probably good
1. Some text editors will automatically change the numbers - i.e. typing enter in github will put `2.` there even if there have been multiple (lazy) `1.` before and/or after
1. Some text editors allow renumbering or even do that automatically (PyCharm), so that would break consistency since we are all mostly using PyCharm.

File: docs/style.md (hunk 8 above)
Comment (Jarno458) [reply docs/style.md:14]: i do prefer double quotes but i hate making it a rule, single quotes have thier use, like if you want to escape a double quote
like 'Received Item:: "{item.name}"'

File: docs/style.md (hunk 9 above)
Comment (LegendaryLinux) [reply docs/style.md:13]: I use a 120 character limit for all my projects, and it's the default in JetBrains products, which we recommend.

**Comment by LegendaryLinux** [comment]
@Ijwu Here's the ESLint config file I use for AginahBot. It represents how I like to see my code. Sometimes I find a rule I don't like and disable or modify it. I've tweaked it for browser usage. I also prefer single-quotes.

```json
{
    "env": {
        "browser": true,
        "commonjs": true,
        "es2021": true
    },
    "extends": "eslint:recommended",
    "parserOptions": {
        "ecmaVersion": "latest"
    },
    "rules": {
        "no-undef": "warn",
        "no-unused-vars": "warn",
        "no-prototype-builtins": "off",
        "no-case-declarations": "off",
        "indent": [
            "error",
            2,
            { "SwitchCase": 1 }
        ],
        "linebreak-style": [
            "error",
            "unix"
        ],
        "quotes": [
            "error",
            "single"
        ],
        "semi": [
            "error",
            "always"
        ]
    }
}
```

**Comment by LegendaryLinux** [comment]
Your analysis of the HTML and CSS is correct, though I would actually prefer different rules for those. I've listed my preferred rules below.

HTML
- Indent should be 2 spaces
- kebab-case for ids and classes

CSS
- Indent should be 2 spaces
- `{` on the same line as the selector
- No space between selector and `{`
  - I know some people don't like this, but I've been doing it so long it'd be like trying to change my breathing rhythm

JS
- Indent should be 2 spaces
- Indent `case` clauses within switch cases with 2 spaces
  - ESLint has an odd syntax for this. `{ "SwitchCase": 1 }` means one indentation, which is defined as two spaces
- Use single quotes
- Semicolons are required after every statement

You can go ahead and include these rules.

File: docs/style.md (hunk 8 above)
Comment (black-sliver) [reply docs/style.md:14]: Python agrees with you on that one
```
>>> "\"test\""
'"test"'
>>> '\'test\''
"'test'"
```
I don't think we have specified what to do in that case, and I doubt a PR gets rejected for `"\"` or `'"'`.

**Review by Ijwu — APPROVED** [review]
2 space indents make me puke. Anyway, PR approved.

**Comment by black-sliver** [comment]
> 2 space indents make me puke.

Also absolutely not my style. But that's something (future) web people will have to fight over :D
